### DISCUSSION TITLE
[DEV] General Features / Bugs Discussion

### DISCUSSION THREAD

**Discussion by MaxwellDPS** [ScanOC/trunk-player#114] [Ideas]
I wanted to make this discussion to talk any features you all want to see or any bugs you want fixed :)

**Comment by bctrainers** [comment]
From personal experience, and from elsewhere over the years...

- Calendar/date-range search of talkgroups and associated transmissions to view historical transmissions.
    - This goes without saying, having these easily linked to via URL would be beneficial. Something like https://somesite.tld/history/?&tg=1234,4321&sdate=01172015&stime=0015&edate=01172015&etime=2300
    - tg = talk groups, singular and multiple.
    - sdate = start date
    - stime = start time
    - edate  = end date
    - etime = end time
    - Date and Time would be in whatever date/time and time zone is set on the Django installation.
    - A sanity limit should be imposed - maybe on the maximum date range limit.
    - Transmissions listed could be considered to play from oldest at top, to newest at the bottom - ignoring pagination.
- Revamp of the userscan page. It currently uses a loop of the REST API to get a listing of all talk groups. This can be incredibly slow to function once the page loads in more than 100 talk groups. Select2 with Redis support would likely speed up this process or caching all talk groups to a single JSON file, generated on a daily basis.
    - pgSQL supports a function known as `json_agg` to generate JSON formatted data. But may become problematic if something other than pgSQL were to be used. A function within models.py could be created and use json_agg somehow to generate a JSON file?
- Allow users to define and save their own scan lists via the user scan page. Would additionally denote a new menu item is added. Editable via their user profile page (would need to be created).
- Expansion of the incident, and scan detail list feature/code.
- Revamp of the password reset, login, and registration pages to support reCaptcha and other Captcha checks. Right now, it has no security checks, and can be brute-forced, and mass-spam-bot registered.
- Centralized template editor in the admin panel for emails dispatched.
- WYSIWYG editor for Web HTML's and Message Popups. I have CKEditor installed on my variant of TrunkPlayer, it works with no problems whatsoever.
- Better control where Message Popups can be displayed at (display on a scan list, web HTMLs, index page, etc), and with a duration/expiration (conditionals) limit.
- Rely on Redis more for key data rather than SQL queries. There are some queries that can sometimes have execution times exceeding 2 to 10 seconds - on a NVMe. Or in the least, rely on caching and better-formatted queries. This may more or less be a limitation of Django and Python.

edit; Formatting/spelling...

**Reply by MaxwellDPS** [reply]
@bctrainers this is great! So inorder here is my plan and where I think these can be integrated.

Forewarning: I am going to put A LOT of focus on rebuilding the webUI, I am learning react atm and have figured out auth and secured the API, now its time for some design, then connecting all the pieces

1. I think a good idea may be an archive page to lookup transmissions. Looks like react can pull URL prams so I would make sure that's how I implement that
2. Agreed the webUI is very slow with lots of ajax, my plan with the new player page is as follows:
   - Implement Infinite scroll vs Pagination 
   - Add transmissions to the ui via websocket, rather than WS triggering a reload
   - Implement A Celery Task Que for Rollup calculations like call metrics and TGs
   - Implement a Metrics and a Call stats Dashboard
   - Looking into a possible [pulse point](https://web.pulsepoint.org/) integration
   - I will look into how to handle the TG issue, sounds like we may just need to implement some indexes
3. A save-able user can list is an excellent idea
4. I like the Incident expansion idea, would love to hear what yall would like to see,
    -  I also Have a desire to bring Mozilla's Deepspeech to the party (Needs some training tho, not sure when I can transcribe 5000 calls lol)
5. All the new logins are going to be JWT based and API based, so I sure can add recaptca.
    - I am also going to lock down all the apis so no un authed viewing of talk groups. (Unless open)
6. On this one I may have to implement a non-django Admin area for this type of management
7. Will likely add this to the non-django admin side for admins, but agreed!
8. I can add message popup limits, as for where they are displayed I will get back to you once I know react a bit better
9. Agreed some of the queries are long, Maybe we can look into adding indexes as those should be a night and day difference. Im hesitant to complicate that stack, but I will address it that way if needed. @bctrainers have you added any indexes to your Tables?


> Two Final thoughts
>  1. This will not be done this year, maybe not even till next fall w/ just me, This will take a bit. 
>  2. I will build out a project on the repo this weekend with bite sized tasks, and hope people better with JS than me hop in from time to time ;)
